File: nim/perfect-numbers/perfect_numbers.nim
import algorithm

type
  Classification* = enum
    Perfect, Abundant, Deficient

iterator factors(n: int): int =
  for i in 1 .. n div 2:
    if n mod i == 0:
      yield i

proc aliquotSum(n: int): int =
  for factor in n.factors:
    result.inc factor

proc classify*(n: int): Classification =
  if n < 1:
    raise newException(ValueError, "Number not positive")

  let sum = n.aliquotSum

  if sum > n:
    Abundant
  elif sum < n:
    Deficient
  else:
    Perfect

File: zig/perfect-numbers/perfect_numbers.zig
pub const Classification = enum {
    deficient,
    perfect,
    abundant,
};

pub fn classify(n: u64) Classification {
    const aliquot_sum = aliquotSum(n);

    if (aliquot_sum < n) return Classification.deficient;
    if (aliquot_sum > n) return Classification.abundant;

    return Classification.perfect;
}

fn aliquotSum(n: u64) u64 {
    var sum: u64 = 0;
    var i: u64 = 1;

    while (i <= n / 2) : (i += 1) {
        if (n % i == 0) sum += i;
    }

    return sum;
}
